Begin implementing Element support for _WKSerializedNode

achristensen07 opened the pull request:
#### 051a5c90bde67c2978dbbf43fd5a83fb731aca48
<pre>
Begin implementing Element support for _WKSerializedNode
<a href="https://rdar.apple.com/156942775">rdar://156942775</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=296587">https://bugs.webkit.org/show_bug.cgi?id=296587</a>

Reviewed by Ryosuke Niwa.

* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::serializeNode const):
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::serializeNode const):
* Source/WebCore/dom/SerializedNode.cpp:
(WebCore::SerializedNode::deserialize):
(WebCore::SerializedNode::QualifiedName::QualifiedName):
(WebCore::SerializedNode::QualifiedName::qualifiedName):
* Source/WebCore/dom/SerializedNode.h:
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::serializeNode const):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/SerializedNode.serialization.in: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm:
(TestWebKitAPI::TEST(SerializedNode, Basic)):

Canonical link: <a href="https://commits.webkit.org/297964@main">https://commits.webkit.org/297964@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d0519b80972a8ff93ba7603fdd96636addb922c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113636 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33343 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23783 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119807 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/64413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115525 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33950 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41913 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86425 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/64413 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116584 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27093 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102117 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66768 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26347 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20237 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/63537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96488 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20331 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123042 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40639 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/30343 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/95277 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41030 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98337 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95031 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40153 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17955 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/36873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18238 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40522 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46035 "Found 1 new failure in dom/SerializedNode.cpp") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/40183 "Build was cancelled. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Running apply-patch; Checked out pull request; Compiled WebKit (cancelled)") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43488 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41938 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Element.h:
```diff
@@ -901,6 +902,9 @@ class Element : public ContainerNode {
 
 private:
+#if ENABLE(ATTACHMENT_ELEMENT)
+    friend class AttachmentAssociatedElement;
```
I don't think we should do this. I'd much rather have `serializeElementWithoutAttributesAndChildren` be publicly accessible than to make this a friend of `Element`.

rniwa commented on Source/WebCore/dom/SerializedNode.cpp:
```diff
@@ -29,16 +29,74 @@
+    constexpr bool createdByParser { false };
+    return WTF::switchOn(WTFMove(element.data), [&] (SerializedElement::Element&& element) -> Ref<WebCore::Element> {
+        return document.createElement(WTFMove(element.name).qualifiedName(), createdByParser);
+    }, [&] (SerializedElement::SliderThumbElement&&) -> Ref<WebCore::Element> {
```
Why do we need to serialize this at all? Anything in UA shadow root we don't need to serialize.

rniwa commented on Source/WebCore/dom/SerializedNode.cpp:
```diff
@@ -29,16 +29,74 @@
+        return document.createElement(WTFMove(element.name).qualifiedName(), createdByParser);
+    }, [&] (SerializedElement::SliderThumbElement&&) -> Ref<WebCore::Element> {
+        return WebCore::SliderThumbElement::create(document);
+    }, [&] (SerializedElement::WebVTTElement&& element) -> Ref<WebCore::Element> {
```
Ditto.

rniwa commented on Source/WebCore/dom/SerializedNode.cpp:
```diff
@@ -29,16 +29,74 @@
+        return WebCore::WebVTTElement::create(element.type, AtomString(WTFMove(element.language)), document);
+    }, [&] (SerializedElement::HTMLImageElement&& element) -> Ref<WebCore::Element> {
+        Ref deserialized = WebCore::HTMLImageElement::create(document);
+        deserializeAssociatedAttachmentElement(WTFMove(element.attachmentAssociatedElement), deserialized, document);
```
I'm pretty sure we don't need to support attachment element.
This feature is only used in Mail's compose window.

rniwa commented on Source/WebCore/dom/SerializedNode.h:
```diff
@@ -36,15 +37,64 @@ class JSValue;
+        SerializedQualifiedName name;
+    };
+    struct SliderThumbElement { };
+    struct WebVTTElement {
```
I'm pretty certain we don't need to serialize these two elements.

rniwa commented on Source/WebCore/dom/SerializedNode.h:
```diff
@@ -36,15 +37,64 @@ class JSValue;
+        bool alreadyStarted;
+    };
+    struct HTMLScriptElement : public ScriptElement { };
+    struct AttachmentAssociatedElement {
```
Or this. We only enable attachment element in Mail's compose window as far as I know.

rniwa commented on Source/WebCore/dom/SerializedNode.h:
```diff
@@ -67,12 +117,12 @@ struct SerializedNode {
         // FIXME: Implement.
     };
-    struct HTMLTemplateElement : public Element {
+    struct HTMLTemplateElement {
```
Why are we making this change??

achristensen07 replied:
The serialization of HTMLTemplateElement is distinct from Element, and we can't make a default-constructed SerializedNode::Element any more.  This is probably temporary until HTMLTemplateElement serialization is implemented

rniwa replied:
I see.

rniwa replied:
Now that I think about this more, this is probably incorrect. A template element can have "regular" child nodes distinct from its template content, and for the purpose of our serialization, we probably want to serialize both, not either.

achristensen07 committed:
Begin implementing Element support for _WKSerializedNode
rdar://156942775
https://bugs.webkit.org/show_bug.cgi?id=296587

Reviewed by Ryosuke Niwa.

* Source/WebCore/dom/Attr.cpp:
(WebCore::Attr::serializeNode const):
* Source/WebCore/dom/ContainerNode.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::serializeNode const):
* Source/WebCore/dom/SerializedNode.cpp:
(WebCore::SerializedNode::deserialize):
(WebCore::SerializedNode::QualifiedName::QualifiedName):
(WebCore::SerializedNode::QualifiedName::qualifiedName):
* Source/WebCore/dom/SerializedNode.h:
* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::serializeNode const):
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Shared/SerializedNode.serialization.in: Added.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm:
(TestWebKitAPI::TEST(SerializedNode, Basic)):

Canonical link: https://commits.webkit.org/297964@main
